File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendNoticeboardRepository.kt
package com.humblesolutions.indsphinx.repository

import android.util.Log
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.SetOptions
import com.humblesolutions.indsphinx.model.Notice
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

class BackendNoticeboardRepository : NoticeboardRepository {
    private val db = FirebaseFirestore.getInstance()

    override fun observeNotices(): Flow<List<Notice>> = callbackFlow {
        val reg = db.collection("NoticeBoard")
            .orderBy("publishedAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                trySend(snapshot?.documents?.mapNotNull { it.toNotice() } ?: emptyList())
            }
        awaitClose { reg.remove() }
    }

    override suspend fun fetchNotice(noticeId: String): Notice? {
        val doc = db.collection("NoticeBoard").document(noticeId).get().await()
        return if (doc.exists()) doc.toNotice() else null
    }

    override fun observeResponseExists(noticeId: String, uid: String): Flow<Boolean> = callbackFlow {
        val reg = db.collection("NoticeBoard").document(noticeId)
            .collection("responses").document(uid)
            .addSnapshotListener { snap, _ -> trySend(snap?.exists() == true) }
        awaitClose { reg.remove() }
    }

    override suspend fun submitResponse(
        noticeId: String,
        uid: String,
        displayName: String,
        flatNo: String,
        recipientType: String,
        selectedOptions: List<String>,
        textResponse: String
    ) {
        val responseRef = db.collection("NoticeBoard").document(noticeId)
            .collection("responses").document(uid)

        val data = mapOf(
            "uid" to uid,
            "displayName" to displayName,
            "flatNo" to flatNo,
            "recipientType" to recipientType,
            "selectedOptions" to selectedOptions,
            "textResponse" to textResponse,
            "submittedAt" to FieldValue.serverTimestamp()
        )
        // The parent `NoticeBoard/{id}` doc is admin-write only per Firestore
        // rules. Earlier this method ran a transaction that incremented the
        // parent's `responseCount` — that write failed with PERMISSION_DENIED,
        // rolled back the whole transaction, and the user saw nothing happen
        // when they tapped Submit. The aggregate is now maintained by the
        // `onNoticeBoardResponseCreated` Cloud Function trigger.
        Log.d(TAG, "submitResponse: noticeId=$noticeId uid=$uid options=${selectedOptions.size} hasText=${textResponse.isNotEmpty()}")
        try {
            responseRef.set(data, SetOptions.merge()).await()
            Log.d(TAG, "submitResponse: success noticeId=$noticeId uid=$uid")
        } catch (e: Exception) {
            Log.e(TAG, "submitResponse: FAILED noticeId=$noticeId uid=$uid", e)
            throw e
        }
    }

    private companion object {
        private const val TAG = "NoticeBoardFlow"
    }

    private fun com.google.firebase.firestore.DocumentSnapshot.toNotice(): Notice? {
        val title = getString("title") ?: return null
        val description = getString("description") ?: ""
        val publishedAt = (get("publishedAt") as? Timestamp)?.toDate()?.time ?: 0L
        val type = getString("type") ?: "normal"
        @Suppress("UNCHECKED_CAST")
        val options = (get("options") as? List<String>) ?: emptyList()
        val allowMultipleChoices = getBoolean("allowMultipleChoices") ?: false
        val responseDeadline = (get("responseDeadline") as? Timestamp)?.toDate()?.time
        val responseCount = (getLong("responseCount") ?: 0L).toInt()
        return Notice(
            id = id,
            title = title,
            description = description,
            publishedAt = publishedAt,
            type = type,
            options = options,
            allowMultipleChoices = allowMultipleChoices,
            responseDeadline = responseDeadline,
            responseCount = responseCount
        )
    }
}

File: iosApp/iosApp/BackendNoticeboardRepository.swift
import Foundation
import FirebaseFirestore

struct Notice: Identifiable, Hashable {
    let id: String
    let title: String
    let description: String
    let publishedAt: Date
    let type: String
    let options: [String]
    let allowMultipleChoices: Bool
    let responseDeadline: Date?
    let responseCount: Int
}

class BackendNoticeboardRepository {
    private let db = Firestore.firestore()

    func observeNotices(onChange: @escaping ([Notice]) -> Void) -> ListenerRegistration {
        return db.collection("NoticeBoard")
            .order(by: "publishedAt", descending: true)
            .addSnapshotListener { snapshot, _ in
                onChange(snapshot?.documents.compactMap { $0.toNotice() } ?? [])
            }
    }

    func fetchNotice(noticeId: String) async throws -> Notice? {
        let doc = try await db.collection("NoticeBoard").document(noticeId).getDocument()
        return doc.exists ? doc.toNotice() : nil
    }

    func observeResponseExists(
        noticeId: String,
        uid: String,
        onChange: @escaping (Bool) -> Void
    ) -> ListenerRegistration {
        return db.collection("NoticeBoard").document(noticeId)
            .collection("responses").document(uid)
            .addSnapshotListener { snap, _ in onChange(snap?.exists == true) }
    }

    func submitResponse(
        noticeId: String,
        uid: String,
        displayName: String,
        flatNo: String,
        recipientType: String,
        selectedOptions: [String],
        textResponse: String
    ) async throws {
        let responseRef = db.collection("NoticeBoard").document(noticeId)
            .collection("responses").document(uid)

        let data: [String: Any] = [
            "uid": uid,
            "displayName": displayName,
            "flatNo": flatNo,
            "recipientType": recipientType,
            "selectedOptions": selectedOptions,
            "textResponse": textResponse,
            "submittedAt": FieldValue.serverTimestamp()
        ]
        // The parent `NoticeBoard/{id}` doc is admin-write only per Firestore
        // rules; an earlier transaction tried to increment its responseCount
        // and failed with PERMISSION_DENIED, swallowing the response write.
        // The counter is now maintained by the
        // `onNoticeBoardResponseCreated` Cloud Function trigger.
        print("[NoticeBoardFlow] submitResponse: noticeId=\(noticeId) uid=\(uid) options=\(selectedOptions.count) hasText=\(!textResponse.isEmpty)")
        do {
            try await responseRef.setData(data, merge: true)
            print("[NoticeBoardFlow] submitResponse: success noticeId=\(noticeId) uid=\(uid)")
        } catch {
            print("[NoticeBoardFlow] submitResponse: FAILED noticeId=\(noticeId) uid=\(uid) error=\(error)")
            throw error
        }
    }
}

private extension DocumentSnapshot {
    func toNotice() -> Notice? {
        guard let data = data(), let title = data["title"] as? String else { return nil }
        let description = data["description"] as? String ?? ""
        let publishedAt = (data["publishedAt"] as? Timestamp)?.dateValue() ?? Date()
        let type = data["type"] as? String ?? "normal"
        let options = data["options"] as? [String] ?? []
        let allowMultipleChoices = data["allowMultipleChoices"] as? Bool ?? false
        let responseDeadline = (data["responseDeadline"] as? Timestamp)?.dateValue()
        let responseCount = data["responseCount"] as? Int ?? 0
        return Notice(
            id: documentID,
            title: title,
            description: description,
            publishedAt: publishedAt,
            type: type,
            options: options,
            allowMultipleChoices: allowMultipleChoices,
            responseDeadline: responseDeadline,
            responseCount: responseCount
        )
    }
}
